### ISSUE TITLE
Command does not execute after CanExecute is false than true

### ISSUE THREAD

**Issue by sam-wheat** [CommunityToolkit/dotnet#322]
### Describe the bug

Command binding for checkbox stops working for unknown reason.

### Steps to reproduce

```text
clone https://github.com/leaderanalytics/LeaderPivot.XAML.WPF.git

set branch to `require_one_measure`

Usage note: Measures are checkboxes in upper left.  Grand Totals is not a measure.  See screenshot below.  

Uncheck than re-check arbitrary measures to verify command handler executes.

Uncheck all measures except last (which will be disabled) as shown in screenshots area below.  

Re-check any measure.  The command does not execute even though it is enabled.  It appears that after CanExecute returns false than true the command does not execute normally.
```

### Expected behavior

Command executes if CanExecute returns true.

### Screenshots

Uncheck all measures except last (which will be disabled):
![Capture](https://user-images.githubusercontent.com/16885210/174929860-da9b2a89-0988-416a-8410-f7ba34c86a67.PNG)

### IDE and version

VS 2022

### Nuget packages

- [ ] CommunityToolkit.Common
- [ ] CommunityToolkit.Diagnostics
- [ ] CommunityToolkit.HighPerformance
- [ ] CommunityToolkit.Mvvm (aka MVVM Toolkit)

### Nuget package version(s)

8.0.0-preview4

### Help us help you

No, just wanted to report this

**Comment by Sergio0694** [comment]
Hey there. Could you please include a minimal repro in the issue we can use. That is, a standalone code snippet that causes the issue, that does not depend on other UI frameworks. That'll help us a lot investigate this. Thanks! 🙂

**Comment by sam-wheat** [comment]
There are maybe thirty lines of code which I have documented for you:


       ToggleMeasureEnabledCommand = new AsyncRelayCommand<Selectable<Measure>>(ToggleMeasureEnabledCommandHandler, (m) => {
    
                // The intent of this logic is to require at least one measure to be selected.
                // Therefore, the checkbox should be enabled if any of the following is true:
                // 1.) Checkbox is unchecked - user should always be able to select (check) a measure.
                // 2.) More than one checkbox is checked.
    
                bool canExecute = !m.IsSelected || ViewBuilder.Measures.Count(x => x.Item.IsEnabled) > 1;
                return canExecute;
            });


      public async Task ToggleMeasureEnabledCommandHandler(Selectable<Measure> measure)
          {
              measure.Item.IsEnabled = measure.IsSelected;
              ToggleMeasureEnabledCommand.NotifyCanExecuteChanged();
              await BuildGrid(null);
          }


      
        <CheckBox 
            Content="{Binding Item.DisplayValue}" 
            IsChecked="{Binding IsSelected }" 
            Command="{Binding RelativeSource={RelativeSource AncestorType=local:LeaderPivotControl}, Path=ToggleMeasureEnabledCommand}" 
            CommandParameter="{Binding}"
            Style="{Binding RelativeSource={RelativeSource AncestorType=local:MeasureContainerCell}, Path=MeasureCheckBoxStyle}" Padding="2" />

**Comment by binarycow** [comment]
@sam-wheat - You are **almost certainly** running into [this issue](https://stackoverflow.com/a/24669638).  If so, it's not an issue with `CommunityToolkit`, it's a problem with WPF.


You can see from the [WPF source code](https://github.com/dotnet/wpf/blob/d4797f1ec6de9347c3725150b5dcacc307859749/src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Controls/Primitives/ButtonBase.cs#L291), that in `ButtonBase`, when the `Command` property changes, the `HookCommand` [method is called](https://github.com/dotnet/wpf/blob/d4797f1ec6de9347c3725150b5dcacc307859749/src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Controls/Primitives/ButtonBase.cs#L309), which updates the `CanExecute` property - that's going to drive whether or not the button is enabled.

[You can also see](https://github.com/dotnet/wpf/blob/d4797f1ec6de9347c3725150b5dcacc307859749/src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Controls/Primitives/ButtonBase.cs#L205) that for `ButtonBase.CommandParameterProperty`, there is **no** `PropertyChangedCallback` defined in its `FrameworkPropertyMetadata` - it simply _does not react_ to a change in command parameter.

----

You can also [see that in the implementation](https://github.com/CommunityToolkit/dotnet/blob/d3250b6f789e86338646c23d72411636f717625d/CommunityToolkit.Mvvm/Input/RelayCommand%7BT%7D.cs#L83) of `RelayCommand<T>` that `CommunityToolkit.Mvvm` provides, if the parameter is `null`, and the default value of `T` is not null - then `false` is returned - your `CanExecute` handler is never executed.

 This includes all value types:

- All `enum` types
- all built-in numeric types (`int`, `double`, etc.), 
- all `struct` (including `DateTime`, `TimeSpan`, `Color`, `Thickness`, etc.), 
- all `record struct`, all `ValueTuple` 
- the list goes on.

It does not include reference types - classes, records (FYI - this includes `string`).  For those, your `CanExecute` handler will receive `null` as its parameter.

----

**IF** you are in fact running into this issue... which, given that you said the `CommandParameter` is a `boolean`, is **almost certainly the case** - you have these three options:

- Create a custom implementation of `IRelayCommand<T>` that will forward `null` command parameters, even if it's a value type.  [Here's some sample code I wrote.](https://gist.github.com/binarycow/d2f5c3132adf2d6b4c583fc5c75da9a5)
    - Con: You can't use the source generator feature in `CommunityToolkit.Mvvm`.
- Use an attached property, [as seen in StackOverflow answer I linked](https://stackoverflow.com/a/24669638), to "refresh" the `Command` databinding when the `CommandParameter` changes.  
    - Con: You have to remember to set `CommandParameter` using the attached property, _instead_ of the regular property, _each and every time_ you have a value type command parameter.
- Subclass `Button`, to do the same thing that ☝️ the attached property does.  [Here's some sample code I wrote.](https://gist.github.com/binarycow/81617a833bf0267386e04d4f0a113cab)
    - Com: You have to remember to use this subclass, instead of using `Button`.

**Comment by binarycow** [comment]
Side note, @sam-wheat - it's not **exactly** the ☝️ above issue.

But, note the important part of what I have outlined above.

**You can also see that for ButtonBase.CommandParameterProperty, there is no PropertyChangedCallback defined in its FrameworkPropertyMetadata - it simply does not react to a change in command parameter.**

In your example code above, you have your `CommandParameter` set via `Binding`.  While the `CommandParameter` does change, `CanExecute` **does not** change automatically.

_but...._ I do see that you appear to have refreshed `CanExecute` manually.... _Hmm...._

Can you give one of the three options I outlined above ☝️ a try?  Maybe the attached property and/or the custom button?

**Comment by sam-wheat** [comment]
@binarycow  Thank for your detailed answer.  I can tell you with confidence that the issue I documented is not related in any way  to [this one](https://stackoverflow.com/questions/335849/wpf-commandparameter-is-null-first-time-canexecute-is-called/24669638#24669638).  Author of cited ticket says: 

> The problem is that the parameter passed to CanExecute is NULL the first time it's called

Nothing about the steps to reproduce, nature, or symptoms of the problem I reported is related to this StackOverflow problem.

You can quickly check this for yourself by following the steps to reproduce which I provided.

Note that you can check than uncheck the boxes to which the command is bound and it works fine.  You can check/uncheck all day every day for 100 years and the control will continue to work fine.  The command executes and the parameter is not null.  It works.

The problem occurs after a certain condition is met:  That condition is that the CanExecute function returns false for any control that it is bound to.  After that, the command does not execute for any bound control - I don't know if the parameter is null or not because I don't get that far.  **The problem is that the control remains enabled and command does not execute.**

May I please suggest that you run the code I have provided to you.  It will only take a moment to clone the repo and see for yourself the sequence of events that leads the problem.

**Comment by Sergio0694** [comment]
Can you provide a small, minimal, platform agnostic, self-contained repro we can use to investigate? 🙂

**Comment by sam-wheat** [comment]
@Sergio0694 Yes as soon as I have a free moment I will provide another example.

**Comment by lassanter** [comment]
@sam-wheat Based on the supplied sample, it appears to me that you are calling NotifyCanExecuteChanged at an inappropriate time. Calling it in the very handler that is being gate kept by the CanExecute that you want to notify comes with the risk of the very behavior you are (were, I assume, given the age of this thread) experiencing. NotifyCanExecuteChanged would be better suited to getting called in response to a property changing. Is there a property that you might use, instead?
